### PR TITLE
Enable Database and Application Access in AWS Terraforms

### PR DESCRIPTION
**Pull request by stevenGravy**
Adds wildcard subdomain to route53 and ACM settings.  
Introduces a variable to enable or disable route53 wildcard subdomain.

### CODE REVIEW

**Comment by stevenGravy** [comment]
Addresses #6819

**Review by smallinsky — COMMENTED** [review]
The code changes looks good to me. @greedy52 Could you also take a look. 

@stevenGravy Have you tested this changes with db access flow ? If not I can give it a try

File: assets/aws/files/bin/teleport-generate-config [hunk 1]
```diff
@@ -81,6 +81,38 @@ write_https_keypairs_section() {
...
+        POSTGRES_PUBLIC_ADDR="'${EXTERNAL_HOSTNAME:-$PUBLIC_IP}:5432'"
+        MONGO_PUBLIC_ADDR="'${EXTERNAL_HOSTNAME:-$PUBLIC_IP}:27017'"
+    fi
+    # set up the database listeners and public addresses
+    cat << EOS >> ${USE_CONFIG_PATH}
+  mysql_listen_addr: "0.0.0.0:3036"
+  postgres_listen_addr: "0.0.0.0:5432"
+  mongo_listen_addr: "0.0.0.0:27017"
+  mysql_public_addr: [${MYSQL_PUBLIC_ADDR}]
+  postgres_public_addr: [${POSTGRES_PUBLIC_ADDR}]
```
Comment (webvictim) [inline assets/aws/files/bin/teleport-generate-config:111]: `postgres_public_addr` is already written elsewhere in the script that generates this config file: https://github.com/gravitational/teleport/blob/master/assets/aws/files/bin/teleport-generate-config#L145-L149

File: assets/aws/files/bin/teleport-generate-config [hunk 2]
```diff
@@ -561,9 +605,6 @@ EOF
     elif [[ "${USE_ACM}" == "true" ]]; then
         echo "ACM support enabled"
 
-        # set up the kubernetes listener
-        write_kubernetes_section TELEPORT_EXTERNAL_HOSTNAME
-
```
Comment (webvictim) [inline assets/aws/files/bin/teleport-generate-config:566]: Why is this removed here?

**Review by webvictim — CHANGES_REQUESTED** [review]
We should have feature flags for each of the database protocols, as existing customers who aren't using them probably don't want extra ports opened on their infrastructure.

You'll want to add some unit tests for this new functionality: https://github.com/gravitational/teleport/tree/master/assets/aws/files/tests

The existing unit tests are also failing: https://github.com/gravitational/teleport/pull/12972/checks?check_run_id=6637577121

```
not ok 84 [ha-proxy-acm] proxy_service.postgres_public_addr is not set
# (in test file assets/aws/files/tests/ha-proxy-acm.bats, line 38)
#   `echo "${PROXY_BLOCK?}" | { ! grep -qE "^  postgres_public_addr: "; }' failed
# proxy_service:
#   enabled: yes
#   listen_addr: 0.0.0.0:3023
#   tunnel_listen_addr: 0.0.0.0:3024
#   web_listen_addr: 0.0.0.0:3080
#   public_addr: gus-tftestkube4.gravitational.io:443
#   ssh_public_addr: gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:3023
#   tunnel_public_addr: gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:3024
#   mysql_listen_addr: "0.0.0.0:3036"
#   postgres_listen_addr: "0.0.0.0:5432"
#   mongo_listen_addr: "0.0.0.0:27017"
#   mysql_public_addr: ['gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:3036']
#   postgres_public_addr: ['gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:5432']
#   mongo_public_addr: ['gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:27017']
#   kubernetes:
#     enabled: yes
#     listen_addr: 0.0.0.0:3026
#     public_addr: ['gus-tftestkube4-proxy-bc9ba568645c3d80.elb.us-east-1.amazonaws.com:3026']
```

File: assets/aws/files/bin/teleport-generate-config (hunk 2 above)
Comment (stevenGravy) [reply assets/aws/files/bin/teleport-generate-config:566]: @webvictim it was already written before the if section

File: assets/aws/files/bin/teleport-generate-config (hunk 1 above)
Comment (stevenGravy) [reply assets/aws/files/bin/teleport-generate-config:111]: I will change the logic to not include

> The code changes looks good to me. @greedy52 Could you also take a look.
> 
> @stevenGravy Have you tested this changes with db access flow ? If not I can give it a try

I have tested them for the ha and starter cluster

File: examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf [hunk 3]
```diff
@@ -72,6 +72,35 @@ resource "aws_security_group_rule" "proxy_ingress_allow_kube" {
   security_group_id = aws_security_group.proxy.id
 }
 
+// Permit inbound to Teleport mysql services
+resource "aws_security_group_rule" "proxy_ingress_allow_mysql" {
+  type              = "ingress"
+  from_port         = 3036
+  to_port           = 3036
+  protocol          = "tcp"
+  cidr_blocks       = ["0.0.0.0/0"]
```
Comment (greedy52) [inline examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf:81]: first time reading these example scripts. assuming we are using `0.0.0.0/0` all other places so i guess this is Ok. 

I think ideally this should be a variable with `["0.0.0.0/0"]` as default so that ppl can customize the CIDR range for better security.

File: examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf [hunk 4]
```diff
@@ -72,6 +72,35 @@ resource "aws_security_group_rule" "proxy_ingress_allow_kube" {
...
+  security_group_id = aws_security_group.proxy.id
```
Comment (greedy52) [inline examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf:82]: nit: maybe add some description in case someone viewing this from AWS console

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 5]
```diff
@@ -61,6 +61,12 @@ variable "route53_domain" {
   type = string
 }
 
+// Whether to set wildcard Route53
+// for use in Application Access
+variable "add_wildcard_route53_record" {
+  type = string
```
Comment (greedy52) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:67]: I think we should use `bool` instead of `string`. `bool` works for `TF_ENV_` too.
```suggestion
  type    = bool
  default = false
```

File: examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf [hunk 6]
```diff
@@ -202,6 +231,63 @@ resource "aws_lb_listener" "proxy_kube" {
   }
 }
 
+// MySQL port
```
Comment (greedy52) [inline examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf:234]: nit: I think just personal preference I would use [`for_each`](https://www.terraform.io/language/meta-arguments/for_each) instead of repeating 3 times for each database port.

**Review by greedy52 — APPROVED** [review]

**Comment by r0mant** [comment]
@webvictim @smallinsky Mind giving this another look guys? Looks like Steven addressed most comments. Thank you!

File: assets/aws/files/bin/teleport-generate-config [hunk 7]
```diff
@@ -81,6 +81,38 @@ write_https_keypairs_section() {
...
+        MONGO_PUBLIC_ADDR="'${EXTERNAL_HOSTNAME:-$PUBLIC_IP}:27017'"
+    fi
+    # set up the database listeners and public addresses
+    cat << EOS >> ${USE_CONFIG_PATH}
+  mysql_listen_addr: "0.0.0.0:3036"
+  postgres_listen_addr: "0.0.0.0:5432"
+  mongo_listen_addr: "0.0.0.0:27017"
+  mysql_public_addr: [${MYSQL_PUBLIC_ADDR}]
+  postgres_public_addr: [${POSTGRES_PUBLIC_ADDR}]
+  mongo_public_addr: [${MONGO_PUBLIC_ADDR}]
```
[the remark this answers is not in the document]
Comment (stevenGravy) [reply assets/aws/files/bin/teleport-generate-config]: @webvictim these were made to be feature flags.

File: examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf (hunk 3 above)
Comment (stevenGravy) [reply examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf]: @greedy52 I've added variables for any references to `0.0.0.0/0`

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 8]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
   default = "false"
 }
 
+//CIDR blocks allows to connect to the bastion SSH port
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:178]:
```suggestion
// CIDR blocks allowed to connect to the bastion SSH port
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 9]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+}
+
+
+//CIDR blocks allowed for egress from bastion
+variable "allowed_bastion_ssh_egress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for ingress for Teleport Proxy ports
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:191]:
```suggestion
// CIDR blocks allowed for ingress for Teleport Proxy ports
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 10]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+variable "allowed_proxy_ingress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for egress from Teleport Proxies
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:197]:
```suggestion
// CIDR blocks allowed for egress from Teleport Proxies
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 11]
```diff
@@ -61,6 +61,34 @@ variable "route53_domain" {
...
 
+// Whether to set wildcard Route53
+// for use in Application Access
+variable "add_wildcard_route53_record" {
+  type    = bool
+  default = false
+}
+
+//whether to enable the mongodb listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:72]:
```suggestion
// whether to enable the mongodb listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 12]
```diff
@@ -61,6 +61,34 @@ variable "route53_domain" {
...
+variable "enable_mongodb_listener" {
+  type    = bool
+  default = false
+}
+
+//whether to enable the mysql listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:79]:
```suggestion
// whether to enable the mysql listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 13]
```diff
@@ -61,6 +61,34 @@ variable "route53_domain" {
...
+variable "enable_mysql_listener" {
+  type    = bool
+  default = false
+}
+
+//whether to enable the postgres listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:86]:
```suggestion
// whether to enable the postgres listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 14]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for ingress for Teleport Monitor ports
+variable "allowed_monitor_ingress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for egress from Teleport Monitor
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:215]:
```suggestion
// CIDR blocks allowed for egress from Teleport Monitor
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 15]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+variable "allowed_monitor_egress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for egress from Teleport Node
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:221]:
```suggestion
// CIDR blocks allowed for egress from Teleport Node
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 16]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+variable "allowed_node_egress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//Internet gateway destination CIDR Block
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:227]:
```suggestion
// Internet gateway destination CIDR Block
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 17]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+variable "internet_gateway_dest_cidr_block" {
+  type = string
+  default = "0.0.0.0/0"
+}
+
+//Route allows for Auth Servers Destination CIDR Block
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:233]:
```suggestion
// Route allowed for Auth Servers Destination CIDR Block
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 18]
```diff
@@ -147,6 +175,73 @@ variable "use_acm" {
...
+variable "auth_aws_route_dest_cidr_block" {
+  type = string
+  default = "0.0.0.0/0"
+}
+
+//Route allows for Node Servers Destination CIDR Block
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:239]:
```suggestion
// Route allowed for Node Servers Destination CIDR Block
```

File: examples/aws/terraform/starter-cluster/vars.tf [hunk 19]
```diff
@@ -31,6 +31,32 @@ variable "route53_domain" {
   type = string
 }
 
+// Whether to add a while a wildcard entry *.proxy.example.com for application access
+variable "add_wildcard_route53_record" {
+  type = string
+}
+
+//whether to enable the mongodb listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/starter-cluster/vars.tf:40]:
```suggestion
// whether to enable the mongodb listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/starter-cluster/vars.tf [hunk 20]
```diff
@@ -31,6 +31,32 @@ variable "route53_domain" {
...
+variable "enable_mongodb_listener" {
+  type    = bool
+  default = false
+}
+
+//whether to enable the mysql listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/starter-cluster/vars.tf:47]:
```suggestion
// whether to enable the mysql listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/starter-cluster/vars.tf [hunk 21]
```diff
@@ -31,6 +31,32 @@ variable "route53_domain" {
...
+variable "enable_mysql_listener" {
+  type    = bool
+  default = false
+}
+
+//whether to enable the postgres listener
+//adds security group setting, maps load balancer to port and adds to teleport config
```
Comment (webvictim) [inline examples/aws/terraform/starter-cluster/vars.tf:54]:
```suggestion
// whether to enable the postgres listener
// adds security group setting, maps load balancer to port and adds to teleport config
```

File: examples/aws/terraform/starter-cluster/vars.tf [hunk 22]
```diff
@@ -59,6 +85,24 @@ variable "use_acm" {
   type = string
 }
 
+//CIDR blocks allows to connect to the SSH port
+variable "allowed_ssh_ingress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for ingress for all Teleport ports
```
Comment (webvictim) [inline examples/aws/terraform/starter-cluster/vars.tf:94]:
```suggestion
// CIDR blocks allowed for ingress for all Teleport ports 
```

File: examples/aws/terraform/starter-cluster/vars.tf [hunk 23]
```diff
@@ -59,6 +85,24 @@ variable "use_acm" {
...
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for ingress for all Teleport ports 
+variable "allowed_ingress_cidr_blocks" {
+  type = list
+  default = ["0.0.0.0/0"]
+}
+
+//CIDR blocks allowed for egress from Teleport
```
Comment (webvictim) [inline examples/aws/terraform/starter-cluster/vars.tf:100]:
```suggestion
// CIDR blocks allowed for egress from Teleport
```

**Review by webvictim — COMMENTED** [review]
Mostly nits from me.

This PR still needs unit tests adding for all the new values which are being set in the config file - this config generation script is really unwieldy and it's very easy to break it unintentionally or introduce syntax errors, which is why I added all the tests in the first place.

You can run the unit tests with `make test-sh` or `make -C build.assets test-sh`. The code lives here: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests

The intent of the tests is to validate that  when the Terraform user-data writes the new values to `/etc/teleport.d/conf`, they're correctly picked up and parsed by the script. You'd probably want to copy this sort of logic and add it to all the relevant test cases: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests/ha-proxy-acm.bats#L41-L51

You should also add a few negative tests to make sure that (for example) `mysql_public_addr` is never written when the feature flag is set to off. Here's an example of how to do a negative test: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests/agent-app.bats#L70-L75

**Comment by stevenGravy** [comment]
> Mostly nits from me.
> 
> This PR still needs unit tests adding for all the new values which are being set in the config file - this config generation script is really unwieldy and it's very easy to break it unintentionally or introduce syntax errors, which is why I added all the tests in the first place.
> 
> You can run the unit tests with `make test-sh` or `make -C build.assets test-sh`. The code lives here: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests
> 
> The intent of the tests is to validate that when the Terraform user-data writes the new values to `/etc/teleport.d/conf`, they're correctly picked up and parsed by the script. You'd probably want to copy this sort of logic and add it to all the relevant test cases: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests/ha-proxy-acm.bats#L41-L51
> 
> You should also add a few negative tests to make sure that (for example) `mysql_public_addr` is never written when the feature flag is set to off. Here's an example of how to do a negative test: https://github.com/gravitational/teleport/blob/master/assets/aws/files/tests/agent-app.bats#L70-L75

I'll write the tests, thanks @webvictim

**Comment by stevenGravy** [comment]
@webvictim the tests have been updated and additional negative tests added for ha and starter proxy.

**Review by webvictim — APPROVED** [review]

**Comment by r0mant** [comment]
@smallinsky Have you had a chance to try this out? Could you approve if it looks good to you?

**Review by smallinsky — APPROVED** [review]
